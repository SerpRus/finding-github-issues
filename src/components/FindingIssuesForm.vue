<template>
  <form
    class="finding-form"
    v-bind:class="classes"
  >
    <label
      for="finding-repository"
      class="finding-form__label"
    >
      <input
        id="finding-repository"
        class="finding-form__input"
        type="text"
        name="finding-input"
        placeholder="vuejs/vue"
        v-model="repositoryName"
      >
    </label>

    <button
      class="finding-form__button"
      type="submit"
      v-on:click="findingBtnHandler"
    >
      Search
    </button>
  </form>

  <PreloaderCircle :isLoader="isLoader"/>

  <FindingErrorMessage
    :isErrorMessage="isErrorMessage"
    :message="message"
  />
</template>

<script>
import axios from 'axios';
import PreloaderCircle from './PreloaderCircle.vue';
import FindingErrorMessage from './FindingErrorMessage.vue';

export default {
  name: 'FindingIssuesForm',
  props: {
    classes: Array,
  },
  components: {
    PreloaderCircle,
    FindingErrorMessage,
  },
  data() {
    return {
      repositoryName: '',
      isLoader: false,
      isErrorMessage: false,
      message: 'Nothing found',
    };
  },
  methods: {
    async findingBtnHandler(e) {
      e.preventDefault();

      if (this.repositoryName === '') return;

      this.isLoader = true;

      const result = await this.fetchIssues();

      this.isLoader = false;

      if (result === null) {
        this.isErrorMessage = true;
      } else {
        this.isErrorMessage = false;
      }
    },
    async fetchIssues() {
      const api = 'https://api.github.com/repos/';
      const url = `${api}${this.repositoryName}/issues`;

      try {
        const { data } = (await axios.get(url));

        return data;
      } catch (err) {
        return null;
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.finding-form {
  display: flex;
  align-content: center;
  justify-content: center;

  &__label {
    display: flex;
  }

  &__input {
    font-size: 20px;

    width: 100%;
    margin-right: 20px;
    padding: 10px 15px;

    border: 2px solid brown;
    border-radius: 8px;

    &:focus {
      border: 2px solid orange;
      outline: none;
    }
  }

  &__button {
    font-size: 20px;

    padding: 10px 15px;

    cursor: pointer;

    border: 2px solid brown;
    border-radius: 8px;
    background-color: #fff;

    &:hover,
    &:focus {
      border: 2px solid orange;
      outline: none;
    }
  }
}

</style>
