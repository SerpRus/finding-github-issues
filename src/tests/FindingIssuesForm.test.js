import { mount } from '@vue/test-utils';
import axios from 'axios';
import FindingIssuesForm from '../components/FindingIssuesForm.vue';

jest.mock('axios');

describe('FindingIssuesForm tests', () => {
  it('Empty request button click', async () => {
    const wrapper = mount(FindingIssuesForm);

    wrapper.find('button').trigger('click');

    expect(axios.get).toHaveBeenCalledTimes(0);
  });

  it('Button click check url with repos name', async () => {
    const wrapper = mount(FindingIssuesForm, {
      props: {
        getIssuesData(data) {
          this.issuesData = data;
        },
      },
    });
    const query = 'vuejs/vue';
    const testData = { data: {} };
    axios.get.mockResolvedValueOnce(testData);

    wrapper.find('input').setValue(query);
    wrapper.find('button').trigger('click');

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith(`https://api.github.com/repos/${query}/issues`);
  });

  it('FindingIssuesForm snapshot', () => {
    const wrapper = mount(FindingIssuesForm);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('FindingIssuesForm isLoader snapshot', () => {
    const wrapper = mount(FindingIssuesForm, {
      data() {
        return {
          isLoader: true,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('FindingIssuesForm isErrorMessage snapshot', () => {
    const wrapper = mount(FindingIssuesForm, {
      data() {
        return {
          isErrorMessage: true,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('FindingIssuesForm isErrorMessage with message snapshot', () => {
    const wrapper = mount(FindingIssuesForm, {
      data() {
        return {
          isErrorMessage: true,
          message: 'test message',
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
