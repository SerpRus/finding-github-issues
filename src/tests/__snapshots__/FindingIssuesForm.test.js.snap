// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindingIssuesForm FindingIssuesForm snapshot 1`] = `
"<form class=\\"finding-form\\"><label for=\\"finding-repository\\" class=\\"finding-form__label\\"><input id=\\"finding-repository\\" class=\\"finding-form__input\\" type=\\"text\\" name=\\"finding-input\\" placeholder=\\"vuejs/vue\\"></label><button class=\\"finding-form__button\\" type=\\"submit\\"> Search </button></form>
<!--v-if-->
<!--v-if-->"
`;
